waiter/src/waiter/fallback.clj

```diff
@@ -0,0 +1,145 @@
+  [descriptor->previous-descriptor search-history-length fallback-state
+   request-time descriptor]
+  (when (-> descriptor :sources :token-sequence seq)
+    (let [{{:keys [token->token-data]} :sources} descriptor
+          service-fallback-period-secs (descriptor->service-fallback-period-secs descriptor)]
+      (when (and (pos? service-fallback-period-secs)
+                 (let [most-recently-modified-token (sd/retrieve-most-recently-modified-token token->token-data)
+                       token-last-update-time (get-in token->token-data [most-recently-modified-token "last-update-time"] 0)]
+                   (->> (t/seconds service-fallback-period-secs)
+                        (t/plus (DateTime. token-last-update-time))
+                        (t/before? request-time))))
+        (loop [iteration 1
+               loop-descriptor descriptor]
+          (when (<= iteration search-history-length)
+            (when-let [previous-descriptor (descriptor->previous-descriptor loop-descriptor)]
+              (let [{:keys [service-id]} previous-descriptor]
+                (if (service-healthy? fallback-state service-id)
+                  (do
+                    (log/info (str "iteration-" iteration) (:service-id descriptor) "falling back to" service-id)
+                    previous-descriptor)
+                  (do
+                    (log/debug (str "iteration-" iteration) "skipping" service-id "as the fallback service"
+                               {:available (service-exists? fallback-state service-id)
+                                :healthy (service-healthy? fallback-state service-id)})
+                    (recur (inc iteration) previous-descriptor)))))))))))
+
+(defn wrap-fallback
+  "Redirects users to a healthy fallback service when the current service has not started or does not have healthy instances."
+  [handler descriptor->previous-descriptor-fn start-new-service-fn assoc-run-as-user-approved?
+   search-history-length fallback-state-atom]
+  (fn wrap-fallback-handler [{:keys [descriptor request-time] :as request}]
+    (let [{:keys [service-id]} descriptor
+          fallback-state @fallback-state-atom]
+      (if (service-healthy? fallback-state service-id)
+        (handler request)
+        (let [auth-user (:authorization/user request)
+              service-approved? (fn service-approved? [service-id] (assoc-run-as-user-approved? request service-id))
+              descriptor->previous-descriptor (fn [descriptor] (descriptor->previous-descriptor-fn service-approved? auth-user descriptor))]
+          (if-let [fallback-descriptor (retrieve-fallback-descriptor
+                                         descriptor->previous-descriptor search-history-length fallback-state request-time descriptor)]
+            (let [fallback-service-id (:service-id fallback-descriptor)
+                  new-handler (middleware/wrap-merge handler {:descriptor fallback-descriptor :fallback-source-id service-id})]
```

Comment:
What if we unconditionally added what the current resolved `service-id` was, instead of only showing when there's a fallback?  

Should we call this `latest-service-id` or `current-service-id`?

Reply:
I like `latest-service-id`, using it.